src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1550 @@
+
+      p_value.sub("first").validate(v => {
+        if (!v)
+          return "First name can't be empty";
+      })
+    }
+
+    export const Person = ({ value } : { value: DialogValue<PersonValue | string> }) => {
+      const val = value.get();
+      if (typeof val == "string")
+          return null;
+
+      const p_value = value.at(val);
+
+      const { country, first, last } = p_value.get();
+      ...
+        <DialogTextInput value={p_value.sub("first")} />
+      ...
+    }
+
+  Note the use of the "value.at()" function to get a handle for a
+  PersonValue that can be used to access the "first" sub value, etc.
+
+ */
+
+import React, { useState } from "react";
+import { useObject, useInit, useOn } from 'hooks';
+import { EventEmitter } from 'cockpit/event';
+
+import cockpit from "cockpit";
+
+import { Button, type ButtonProps } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { FormGroup, type FormGroupProps, FormHelperText } from "@patternfly/react-core/dist/esm/components/Form";
+import { TextInput, type TextInputProps } from "@patternfly/react-core/dist/esm/components/TextInput";
+import { Alert } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import {
+    HelperText, HelperTextItem, type HelperTextItemProps
+} from "@patternfly/react-core/dist/esm/components/HelperText";
+import { Checkbox } from "@patternfly/react-core/dist/esm/components/Checkbox";
+import {
+    FormSelect, FormSelectOption, type FormSelectProps,
+} from "@patternfly/react-core/dist/esm/components/FormSelect";
+import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
+
+const _ = cockpit.gettext;
+
+function debug(...args: unknown[]) {
+    if (window.debugging == "all" || window.debugging?.includes("dialog"))
+        console.debug("dialog:", ...args);
+}
+
+type ArrayElement<ArrayType> =
+  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
+
+function toSpliced<T>(arr: T[], start: number, deleteCount: number, ...rest: T[]): T[] {
+    const copy = [...arr];
+    copy.splice(start, deleteCount, ...rest);
+    return copy;
+}
+
+export class DialogValue<T> {
+    /* eslint-disable no-use-before-define */
+    #dialog: DialogState<unknown>;
+    /* eslint-enable */
+    #getter: () => T;
+    #setter: (val: T, is_aux: boolean) => void;
+    #path: string;
+
+    constructor(
+        dialog: DialogState<unknown>,
+        getter: () => T,
+        setter: (val: T, is_aux: boolean) => void,
+        path: string
+    ) {
+        this.#dialog = dialog;
+        this.#getter = getter;
+        this.#setter = setter;
+        this.#path = path;
+    }
+
+    validation_text(): string | undefined {
+        return this.#dialog._get_validation(this.#path);
+    }
+
+    get(): T {
+        return this.#getter();
+    }
+
+    set(val: T): void {
+        this.#setter(val, false);
+    }
+
+    set_aux(val: T): void {
+        this.#setter(val, true);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-0ef97292-20251218-063518-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)